### PR TITLE
Produce shorter SQL for `filter()` after `mutate()`

### LINKED ISSUE #859: select + filter generates unecessary (?) subquery?
``` r
library(dbplyr)
library(dplyr, warn.conflicts = FALSE)

lazy_frame(x = 1, y = 1) |> 
  select(y) |> 
  filter(y > 1000)
#> <SQL>
#> SELECT *
#> FROM (
#>   SELECT `y`
#>   FROM `df`
#> ) `q01`
#> WHERE (`y` > 1000.0)
```

<sup>Created on 2022-05-05 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### PR DESCRIPTION
**Pull request by mgirlich**
Closes #859.

`mutate()` + `filter()`

```r
lf <- lazy_frame(x = 1, y = 2)
lf %>% 
  mutate(x = x + 1) %>% 
  filter(y == 1)
```

Before

```sql
SELECT *
FROM (
  SELECT `x` + 1.0 AS `x`, `y`
  FROM `df`
) `q01`
WHERE (`y` = 1.0)
```

After

```sql
SELECT `x` + 1.0 AS `x`, `y`
FROM `df`
WHERE (`y` = 1.0)
```

`filter()` + `filter()`

```r
lf <- lazy_frame(x = 1, y = 2)
lf %>% 
  filter(y == 1) %>% 
  filter(x == 1)
```

Before

```sql
SELECT *
FROM (
  SELECT *
  FROM `df`
  WHERE (`y` = 1.0)
) `q01`
WHERE (`x` = 1.0)
```

After

```sql
SELECT *
FROM `df`
WHERE (`y` = 1.0) AND (`x` = 1.0)
```

### CODE REVIEW

File: tests/testthat/test-verb-filter.R [hunk 1]
```diff
@@ -46,6 +62,61 @@ test_that(".preserve is not supported", {
   expect_snapshot(error = TRUE, lf %>% filter(x == 1, .preserve = TRUE))
 })
 
+test_that("filter() works with mutate()", {
+  lf <- lazy_frame(x = 1, y = 2)
+  named_list <- function(...) {
+    out <- list(...)
+    if (is_null(names(out))) {
+      names(out) <- names2(out)
+    }
```
Comment (hadley) [inline tests/testthat/test-verb-filter.R:71]: What's this doing/why is it needed? I'm guessing it's something to do with `names(x)` being `NULL` vs `""`?

**Review by hadley — COMMENTED** [review]
Nice improvement!

**Comment by hadley** [comment]
This also fixes a small buglet:

```R
library(dbplyr)

lazy_frame(x = 1, y = 1) |> 
  select(y) |> 
  filter(z > 1000)
```

currently generates the incorrect:

```sql
SELECT *
FROM (
  SELECT `y`
  FROM `df`
) `q01`
WHERE (`z` > 1000.0)
```

But after this PR it generates

```sql
SELECT `y`
FROM `df`
WHERE (`z` > 1000.0)
```

Might be worth adding a test for this?

**Comment by mgirlich** [comment]
> This also fixes a small buglet:
> 
> ```r
> library(dbplyr)
> 
> lazy_frame(x = 1, y = 1) |> 
>   select(y) |> 
>   filter(z > 1000)
> ```
> 
> currently generates the incorrect:
> 
> ```sql
> SELECT *
> FROM (
>   SELECT `y`
>   FROM `df`
> ) `q01`
> WHERE (`z` > 1000.0)
> ```

Yeah, I like how it now generates the correct error 😉 

``` r
library(dbplyr)
library(dplyr, warn.conflicts = FALSE)

lazy_frame(x = 1, y = 1) |> 
  select(y) |> 
  filter(z > 1000)
#> Error in `filter()`:
#> ! Problem while computing `..1 = z > 1000`
#> Caused by error:
#> ! Object `z` not found.
```

<sup>Created on 2022-07-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

I also added another test for your fixed example.

**Review by hadley — APPROVED** [review]

**Pull request merged**